### PR TITLE
Add graph rewrite primitives

### PR DESCRIPTION
**Pull request by staticfloat**
Greetings, Keras devs.

I asked some questions about graph rewriting on the slack channel a while back, and the feedback I got from there has led me to this pull request, implementing what I have come to believe is the most fundamental operation required to perform "net surgery" within Keras.  My use case stems from wanting to build a library of network optimizations (things like factorizing a square-ish `Dense` layer into two smaller `Dense` layers using SVD and restricting the rank of the SVD representation) and needing the machinery available to replace subgraphs within a model with a new architecture and new weights.  I started by looking at how I might insert nodes, but quickly discovered that although Keras provides easy methods to grow a model "downstream", it is difficult to remove nodes from the "upstream" of a Model, or to, more ambitiously, "rebase" or "graft" a portion of a Model onto another one.  This pull request provides a prototype implementation of this "rebasing", allowing a section of a predefined `Model` to be rebased on top of another `Model`.  This allows for simple and flexible model manipulations; inserting layers and removing layers can all be phrased within the framework of adding layers onto the end of a model, and rebasing a model onto a new base model.

This PR adds four methods: the fundamental `rebase()` operation, a slightly higher level `switch_fork()` operation, and the very high-level `insert_layer()` and `remove_layer()` operations.  I have provided tests for all four methods.

- `rebase(model, onto, rebase_from)` rebases argument `model` onto argument `onto`, both of which are of type `Model`/`Container`.  The "rebase" operation allows one to completely rewrite a graph on top of another graph, and the implementation of `rebase()` performs the following steps:
    -  It first identifies the "upstream edge" of the model it will be reconstructing.  By default, this is `model.inputs`, but if we are, for instance, replacing the first half of `model` with a new graph, then this upstream edge (passed in as `rebase_from`) must be realized, and correspondences between these layers must be drawn from the old model to the new model.
    - It next steps through the old model, recreating nodes from the old model on top of the new model, with the same configuration (and weights, if possible) as specified within the old model.
    - Finally, a new `Model`/`Container` is returned with the desired topology

- `switch_fork(model, old_prong, new_prong)` uses `rebase()` to effect a common mid-level operation; as a user, I wish to replace a particular linear subgraph within a `Model`, so I construct the new subgraph I wish to use as a branch extending off the side of the original model.  `switch_fork()` can then take everything downstream from the linear subgraph of interest and place it onto the fork that I just created. Example, taken from the `switch_fork()` docstring:
```
`model` contains the following topology:

            [A (Input)] -> [B] -> [C] -> [D] -> [E] -> [F (Output)]

        You then create a "fork", by creating two new layers off of `B`:

                              /-> [C] -> [D] -> [E] -> [F (Output)]
            [A (Input)] -> [B]
                              \-> [G] -> [H]

        By calling `switch_fork(model, D, H)`, the model will transform to:

                              /-> [C] -> [D]
            [A (Input)] -> [B]
                              \-> [G] -> [H] -> [E] -> [F (Output)]

        And the old prong can be safely ignored, resulting in a topology of:

            [A (Input)] -> [B] -> [G] -> [H] -> [E] -> [F (Output)]
```
- `insert_layer(model, after, to_insert)` uses `switch_fork()` to insert layers directly after another layer within a model.
- `remove_layer(model, to_remove)` uses `switch_fork()` to remove layers from a `Model`, throwing `ValueError`s when attempting to remove nonsensical Layers such as `Input` or `Merge` layers.

I am very open to feedback on everything from the technical details of how the `rebase()` function explores the Keras node graph, to the naming/placement of the functions.  Strictly speaking, none of this touches the internals of Keras itself, so it could be that this belongs in [keras-contrib](https://github.com/farizrahman4u/keras-contrib), but I was not explicitly told to open this PR there versus here, so I figured I'd start with here and see which way the wind blew.  The actual use cases of this work I hope to demo later, so for now it's just abstract capabilities rather than concrete applications.

Questions I have about the current implementation:
 - [ ] At the moment, I am very cavalier about things like "layers with multiple input nodes"; I pretty much assume that everything will have a single input node that can contain one or more inbound layers.  Where can I find a Layer that has multiple input nodes so I can see how horribly this breaks?
 - [ ] My layer recreation code is also rather cavalier about [creation of nodes with multiple inputs](https://github.com/staticfloat/keras/blob/4d57365b40875b4d167507a1346fecd21c9f3814/keras/engine/topology.py#L2940-L2945), is there a more principled way to feed inputs into Layer creation than `LayerName(*layer_deps)`?

### CODE REVIEW

**Comment by ahundt** [comment]
Graph rewiring is important for several applications! FYI, you're almost certainly going to need to follow the `API Design Review` process explained in [CONTRIBUTING.md](https://github.com/fchollet/keras/blob/master/CONTRIBUTING.md#pull-requests).

**Comment by fchollet** [comment]
Currently, there is identity between the graph of layers described by a Keras model, and the "TensorFlow graph" underlying it. This correspondence is important to avoid bugs and avoid user confusion.

TensorFlow graphs do not support graph rewiring. So what approach are you taking?

Also, could you describe a few practical use cases?

**Comment by ahundt** [comment]
> Also, could you describe a few practical use cases?

I don't know about what's specifically implemented in this PR, but here are three in the general case:

1. Ensuring model.fit can take input tensors as parameters for the "ideal" api in #7102
2. Support for both training and validation input tensors for 1.
3. Rewiring an  imagenet pretrained network's graph weights to become a segmentation network.
4. I've had brief discussions with others where they said they switched from tensorflow to pytorch because graph rewiring was easier for reinforcement learning (sorry that one is vague). 

I'm interested to hear from @staticfloat if any of the above can be supported with the changes here. Could you also make a small (~100 lines) canonical example program demonstrating usage?

A really great example would be transforming an [imagenet vgg16 to segmentation vgg16 like this example](https://github.com/aurora95/Keras-FCN/blob/master/utils/transfer_FCN.py#L19) without going to disk or regenerating the network in different modes.

**Comment by staticfloat** [comment]
> Currently, there is identity between the graph of layers described by a Keras model, and the "TensorFlow graph" underlying it. This correspondence is important to avoid bugs and avoid user confusion.
> TensorFlow graphs do not support graph rewiring. So what approach are you taking?

I am taking inspiration from the `Sequential` model's `pop()` functionality; a Keras model doesn't care if there are "forks" in the model with branches that extend off to the sides and are not reachable from the outputs.  Concretely, Let's say I have a `Sequential` model named `M` that looks like this:

```
M:
    [A] -> [B] -> [C] -> [D]
```

If I call `M.pop(); M.pop(); M.add(E)`, to the user, it looks like this now exists:

```
M:
    [A] -> [B] -> [E]
```

But what actually exists is this:

```
M:
                  /-> [E (output)]
    [A] -> [B]
                  \-> [C] -> [D]
```

The `[C] -> [D]` branch is not listed within `M's` outputs, and therefore will not effect evaluation, and so the `pop()` operation need not worry about them.  My approach is similar; grow new graphs on top of existing ones where possible.  I have not done anything to the Keras internals here, this is all just `Layer`-level manipulations of the graph, with the exception that I make use of `x._keras_history` and introspect to walk the graph of `Layer` objects in order to grow these graphs.

> Also, could you describe a few practical use cases?

Yes, I'd be happy to.  I am developing a library of network optimization passes, including, among others, Tensor Factorization as described in [Kim 2015](https://arxiv.org/abs/1511.06530).  This is a technique to factorize layers such as Dense layers or Conv2D layers into multiple, smaller, layers that have less weights and connections and therefore executes faster, but with a slight accuracy penalty.  Without going into too much detail, you could imagine factorizing a matrix into three matrices using SVD.  If I have a matrix `W`, using SVD I can factorize it into the typical `U S V'` form, and I can then start eliminating computational work by restricting the "rank" of my SVD representation, e.g. throwing away the eigenvalues with smallest magnitude.  This will cause the `U` and `V` matrices to be rectangular, and `S` diagonal matrix to shrink.  To relate this back to neural networks, I can easily define my new rectangular `U` matrix as a `Dense` layer, and combined `S V'` into another `Dense` layer.  Assuming that I restrict the rank enough, this can have significant computational savings.

This can be extended to tensors of higher rank than just matrices, turning one large `Conv2D` layer into three smaller ones, for instance.  My use case is, given a pre-trained model, I inspect the model for promising optimization opportunities to speed up evaluation as much as possible.  One of my optimizations is this tensor factorization, and so I "replace" things like `Dense` layers and `Conv2D` layers with the appropriate factorized weights and the new architecture to replace the old layers.  There are other optimizations possible, but this one is a convenient case study that is straightforward so let's just talk about this one.

I hope this helps put my original message in perspective.  I am primarily concerned with the ability to "replace" sections of a graph.  Most of my applications are very simple (e.g. "replace a single Conv2D with three Conv2D layers") but tools like the ones developed in this PR make it much, much easier to automatically apply these optimizations in a principled way to a Model than it would be otherwise to manually create the new graph every time.

> A really great example would be transforming an [imagenet vgg16 to segmentation vgg16 like this example](https://github.com/aurora95/Keras-FCN/blob/master/utils/transfer_FCN.py#L19) without going to disk or regenerating the network in different modes.

Looking at that example, it looks like all that's happening is that some `Dense` weights are being reshaped and then applied to new `Conv2D` layers.  I don't think you need this functionality for that kind of an operation.  If you refer to my network diagram above, you can think of your operation as setting `[E]` as the new `Conv2D` layers, and `[C] -> [D]` as the old `Dense` layers.  All you need to do is create the new `Conv2D` layers as a new "branch" growing off of your model, then create a new `Model` object that has your new branch set as the `outputs` instead of the old branch.  Example:

```python
from keras.applications.vgg16 import *
from keras.layers import Conv2D
from numpy import reshape

vgg16 = VGG16()
# Get the nodes we're interested in from vgg16
fork_base = vgg16.get_layer("block5_pool")
fc1 = vgg16.get_layer("fc1")
fc2 = vgg16.get_layer("fc2")
predictions = vgg16.get_layer("predictions")

# Get the weights we're interested in and reshape them
fc1_weights = fc1.get_weights()
fc1_weights[0] = reshape(fc1_weights[0], (7,7,512,4096))
fc2_weights = fc2.get_weights()
fc2_weights[0] = reshape(fc2_weights[0], (1,1,4096,4096))
predictions_weights = predictions.get_weights()
predictions_weights[0] = reshape(predictions_weights[0], (1,1,4096,1000))

# Build our new fork
x = Conv2D(4096, 7, 7, activation='relu', border_mode='same', name='fc1_replacement', weights=fc1_weights)(fork_base)
x = Conv2D(4096, 1, 1, activation='relu', border_mode='same', name='fc2_replacement', weights=fc2_weights)(x)
x = Conv2D(1000, 1, 1, activation='linear', name='predictions_replacement', weights=predictions_weights)(x)

# Wrap this up in a new Model:
new_model = Model(inputs=vgg16.inputs, outputs=[x], loss=vgg16.loss, optimizer=vgg16.optimizer, metrics=vgg16.metrics)
```

You can, of course, use the methods in this PR to replace the `Dense` layers, but it will do the exact same thing as what I just did above manually.  This because more important when there are layers downstream of the layers you want to edit that must be recreated.  That's what this PR is really useful for, and that's the whole point of the `rebase()` code.

**Comment by ahundt** [comment]
@staticfloat I linked to the wrong thing, sorry! The [resnet example](https://github.com/aurora95/Keras-FCN/blob/master/models.py#L190) and [conversion code](https://github.com/aurora95/Keras-FCN/blob/master/utils/transfer_FCN.py#L86) is clearer. What happens is Max pooling layers with a stride of `(2,2) are dropped, and the input to the next convolution is dilated at a rate of `(2,2)` to keep the input and output dimensions constant. That's where actual rewiring is most relevant.

**Comment by fchollet** [comment]
I think this looks very interesting, but it's also a lot of experimental features and code, and it's something that's not on the roadmap. My recommendation would be to put in contrib, so that users can start playing with it, and when use cases are clear and in-demand (and when the implementation is battle-tested), we will consider adding it core. Since it's an add-on, that shouldn't be an issue.

**Comment by ahundt** [comment]
I got to look a bit closer and it seems nice! Rebase looks like it might be perfect for connecting new inputs. One question, what sort of tensor/dimension checks are there to ensure the user doesn't go wrong?

> My recommendation would be to put in contrib

Oh yeah, good thinking. It can really get tried out and mature a bit there, then be moved in once it is hardened. https://github.com/farizrahman4u/keras-contrib

**Comment by staticfloat** [comment]
Great, thanks for the feedback guys.  I will open a new PR to keras-contrib.

**Comment by ahundt** [comment]
@staticfloat did you get a chance to submit this to keras-contrib?
